tools/device.c

```diff
@@ -137,43 +143,114 @@ static int bl_device__is_ttyACM(const struct dirent *entry)
 	return strncmp(entry->d_name, "ttyACM", 6) == 0;
 }
 
-static int bl_device__scan(void)
+static bool bl_device__list_append(
+		struct bl_device **devices,
+		unsigned *count,
+		const char *device_path,
+		const char *device_serial)
+{
+	struct bl_device *temp;
+
+	temp = realloc(*devices, sizeof(**devices) * (*count + 1));
+	if (temp == NULL) {
+		return false;
+	}
+	*devices = temp;
+
+	temp[*count].device_path = strdup(device_path);
+	temp[*count].device_serial = strdup(device_serial);
+
+	if (temp[*count].device_path   == NULL ||
+	    temp[*count].device_serial == NULL) {
+		free(temp[*count].device_path);
+		free(temp[*count].device_serial);
+		return false;
+	}
+
+	*count = *count + 1;
+	return true;
+}
+
+void bl_device_list_free(
+		struct bl_device *devices,
+		unsigned count)
+{
+	for (unsigned i = 0; i < count; i++) {
+		free(devices[i].device_path);
+		free(devices[i].device_serial);
+	}
+
+	free(devices);
+}
+
+bool bl_device_list_get(
+		struct bl_device **devices_out,
+		unsigned *count_out)
 {
+	struct bl_device *list = NULL;
 	struct dirent **namelist;
-	int n, found = 0;
+	unsigned found = 0;
+	bool res = false;
+	int count;
 
-	n = scandir("/dev/", &namelist, bl_device__is_ttyACM, NULL);
-	if (n == -1) {
+	count = scandir("/dev/", &namelist, bl_device__is_ttyACM, NULL);
+	if (count == -1) {
 		fprintf(stderr, "scandir failed: %s\n", strerror(errno));
-		exit(EXIT_FAILURE);
+		return false;
 	}
 
-	while (n--) {
-		if (bl_device__match(namelist[n]->d_name)) {
+	for (int i = 0; i < count; i++) {
+		char *serial;
+		if (bl_device__match(namelist[i]->d_name, &serial)) {
 			int ret = snprintf(dev_node, sizeof(dev_node) - 1,
-					"/dev/%s", namelist[n]->d_name);
+					"/dev/%s", namelist[i]->d_name);
 			if (ret < 0 || ret >= (int)(sizeof(dev_node) - 1)) {
 				fprintf(stderr, "Error saving device path: %s\n",
-						namelist[n]->d_name);
-				free(namelist[n]);
-				break;
+						namelist[i]->d_name);
+				free(serial);
+				goto out;
 			}
-			found++;
+
+			if (!bl_device__list_append(&list, &found,
+					dev_node, serial)) {
+				free(serial);
+				goto out;
+			}
+			free(serial);
 		}
-		free(namelist[n]);
 	}
-	free(namelist);
 
-	return found;
+	if (found) {
+		*devices_out = list;
+		*count_out = found;
+		res = true;
+	}
+
+out:
+	if (res == false) {
+		bl_device_list_free(list, found);
+	}
+	for (int i = 0; i < count; i++) {
+		free(namelist[i]);
+	}
+	free(namelist);
+	return res;
 }
 
 void get_dev(int dev, char *argv[])
 {
 	if ((strncmp(argv[dev], "auto", 4) ||
```

Comment:
Should we support all possible invocations? Since "auto" could be a file path, maybe we should have it always be a flag? But if it's a flag we'd have to stop it being positional.

Reply:
In a later patch I move this out of the device code into the bl.c code.

So this function is gone from the library functionality, and

```c
int bl_device_open(const char *dev_path);
```

opens the first found bloodlight device if given a NULL path.

So the argument handling is now purely an issue for bl.c.

Reply:
Removed the `auto` option as a new commit.